### ISSUE TITLE
No support for extrapolating XDR data from transaction

### ISSUE THREAD

**Issue by vitchor** [StellarCN/py-stellar-base#223]
It would be really useful to be able to extrapolate XDR information from transactions "result_meta_xdr".

**Comment by overcat** [comment]
We have not provided a tool to handle it yet, but you can get the extrapolate it like this. 
```
# xdr: https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAQAAAAIAAAADAYu3FQAAAAAAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAABkRxSUqgFiuIoACFAIAAABvAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAC5mFTKkAAAAC66PM5MAAAAAAAAAAAAAAAEBi7cVAAAAAAAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAGRHFJSqAWK4igAIUAkAAAG8AAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAALmYVMqQAAAALro8zkwAAAAAAAAAAAAAAAQAAAAYAAAADAYu2xwAAAAIAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAAAABujbrAAAAAAAAAABU0hYAAAAAADlOMj3OZP8orENyYJ2H%2BXpUVJEUQMf0EclKEV%2FTfo%2BpAAAAAEKdphTAJiWgAABDYkAAAAAAAAAAAAAAAAAAAACAAAAAgAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAAAG6NusAAAAAwGLtscAAAABAAAAABlOzgWjBmqrEHvGIemKVSliLlABFNbTkSwFYsohekGmAAAAAVNIWAAAAAAA5TjI9zmT%2FKKxDcmCdh%2Fl6VFSRFEDH9BHJShFf036PqQAAAAAAAAASn%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FAAAAAQAAAAEAAACW2U3qPQAAAAAAAAAAAAAAAAAAAAAAAAABAYu3FQAAAAEAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAAABU0hYAAAAAADlOMj3OZP8orENyYJ2H%2BXpUVJEUQMf0EclKEV%2FTfo%2BpAAAAAAAAABKf%2F%2F%2F%2F%2F%2F%2F%2F%2F8AAAABAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMBi7cVAAAAAAAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAGRHFJSqAWK4igAIUAkAAAG8AAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAALmYVMqQAAAALro8zkwAAAAAAAAAAAAAAAQGLtxUAAAAAAAAAABlOzgWjBmqrEHvGIemKVSliLlABFNbTkSwFYsohekGmAAAAZEcUlKoBYriKAAhQCQAAAbsAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAuZhUypAAAAAqkGJtAAAAAAAAAAAA%3D&type=TransactionMeta
import base64

from stellar_base import stellarxdr

xdr = "AAAAAQAAAAIAAAADAYu3CgAAAAAAAAAA+Qf/14y+eUICTL6t29QysQh1lq8e4Y9ShDwbtkCnYDIAAAAAPp7KCAFLW0EAAhCoAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAABAYu3CgAAAAAAAAAA+Qf/14y+eUICTL6t29QysQh1lq8e4Y9ShDwbtkCnYDIAAAAAPp7KCAFLW0EAAhCpAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAACAAAABAAAAAMBijziAAAAAQAAAAAfyCUly0wwjW/8GArCBkffwuVxr1PVOFfyMR6DsV8g7wAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAC3gWa1//////////wAAAAEAAAAAAAAAAAAAAAEBi7cKAAAAAQAAAAAfyCUly0wwjW/8GArCBkffwuVxr1PVOFfyMR6DsV8g7wAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAGl7I61//////////wAAAAEAAAAAAAAAAAAAAAMBi6NeAAAAAQAAAABxeNaH6pI9wQS75bqNZwbRslttfUI3XO2YBylykuDE8QAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAII66al//////////wAAAAEAAAAAAAAAAAAAAAEBi7cKAAAAAQAAAABxeNaH6pI9wQS75bqNZwbRslttfUI3XO2YBylykuDE8QAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAEagH6l//////////wAAAAEAAAAAAAAAAAAAAAQAAAADAYu3BwAAAAEAAAAA90xENER+IjZu+C6NhKKj83zqYCi1HtYCORsV9opylfAAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAYTcoG3f/////////8AAAABAAAAAAAAAAAAAAABAYu3CgAAAAEAAAAA90xENER+IjZu+C6NhKKj83zqYCi1HtYCORsV9opylfAAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAYTgRqif/////////8AAAABAAAAAAAAAAAAAAADAYu24wAAAAEAAAAAmlZqRKk9qqEpuyNkDV4LQNI4pBqpW7VR400/bNs2DPUAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAADb1jQf/////////8AAAABAAAAAAAAAAAAAAABAYu3CgAAAAEAAAAAmlZqRKk9qqEpuyNkDV4LQNI4pBqpW7VR400/bNs2DPUAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAADYL/lf/////////8AAAABAAAAAAAAAAA="
xdr_decode = base64.b64decode(xdr)
unpacker = stellarxdr.Xdr.StellarXDRUnpacker(xdr_decode)
transaction_meta = unpacker.unpack_TransactionMeta()
```
This part of the code does need to be optimized.

**Comment by vitchor** [comment]
Ok, @overcat – thanks!

**Comment by AnimusXCASH** [comment]
> We have not provided a tool to handle it yet, but you can get the extrapolate it like this.
> 
> ```
> # xdr: https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAQAAAAIAAAADAYu3FQAAAAAAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAABkRxSUqgFiuIoACFAIAAABvAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAC5mFTKkAAAAC66PM5MAAAAAAAAAAAAAAAEBi7cVAAAAAAAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAGRHFJSqAWK4igAIUAkAAAG8AAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAALmYVMqQAAAALro8zkwAAAAAAAAAAAAAAAQAAAAYAAAADAYu2xwAAAAIAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAAAABujbrAAAAAAAAAABU0hYAAAAAADlOMj3OZP8orENyYJ2H%2BXpUVJEUQMf0EclKEV%2FTfo%2BpAAAAAEKdphTAJiWgAABDYkAAAAAAAAAAAAAAAAAAAACAAAAAgAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAAAG6NusAAAAAwGLtscAAAABAAAAABlOzgWjBmqrEHvGIemKVSliLlABFNbTkSwFYsohekGmAAAAAVNIWAAAAAAA5TjI9zmT%2FKKxDcmCdh%2Fl6VFSRFEDH9BHJShFf036PqQAAAAAAAAASn%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FAAAAAQAAAAEAAACW2U3qPQAAAAAAAAAAAAAAAAAAAAAAAAABAYu3FQAAAAEAAAAAGU7OBaMGaqsQe8Yh6YpVKWIuUAEU1tORLAViyiF6QaYAAAABU0hYAAAAAADlOMj3OZP8orENyYJ2H%2BXpUVJEUQMf0EclKEV%2FTfo%2BpAAAAAAAAABKf%2F%2F%2F%2F%2F%2F%2F%2F%2F8AAAABAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMBi7cVAAAAAAAAAAAZTs4FowZqqxB7xiHpilUpYi5QARTW05EsBWLKIXpBpgAAAGRHFJSqAWK4igAIUAkAAAG8AAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAALmYVMqQAAAALro8zkwAAAAAAAAAAAAAAAQGLtxUAAAAAAAAAABlOzgWjBmqrEHvGIemKVSliLlABFNbTkSwFYsohekGmAAAAZEcUlKoBYriKAAhQCQAAAbsAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAuZhUypAAAAAqkGJtAAAAAAAAAAAA%3D&type=TransactionMeta
> import base64
> 
> from stellar_base import stellarxdr
> 
> xdr = "AAAAAQAAAAIAAAADAYu3CgAAAAAAAAAA+Qf/14y+eUICTL6t29QysQh1lq8e4Y9ShDwbtkCnYDIAAAAAPp7KCAFLW0EAAhCoAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAABAYu3CgAAAAAAAAAA+Qf/14y+eUICTL6t29QysQh1lq8e4Y9ShDwbtkCnYDIAAAAAPp7KCAFLW0EAAhCpAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAACAAAABAAAAAMBijziAAAAAQAAAAAfyCUly0wwjW/8GArCBkffwuVxr1PVOFfyMR6DsV8g7wAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAC3gWa1//////////wAAAAEAAAAAAAAAAAAAAAEBi7cKAAAAAQAAAAAfyCUly0wwjW/8GArCBkffwuVxr1PVOFfyMR6DsV8g7wAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAGl7I61//////////wAAAAEAAAAAAAAAAAAAAAMBi6NeAAAAAQAAAABxeNaH6pI9wQS75bqNZwbRslttfUI3XO2YBylykuDE8QAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAII66al//////////wAAAAEAAAAAAAAAAAAAAAEBi7cKAAAAAQAAAABxeNaH6pI9wQS75bqNZwbRslttfUI3XO2YBylykuDE8QAAAAFUWFQAAAAAABUR3CcGpzdhOYIBD+Ry73r9ciDEaUzam1HktDssrXiEAAAAAEagH6l//////////wAAAAEAAAAAAAAAAAAAAAQAAAADAYu3BwAAAAEAAAAA90xENER+IjZu+C6NhKKj83zqYCi1HtYCORsV9opylfAAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAYTcoG3f/////////8AAAABAAAAAAAAAAAAAAABAYu3CgAAAAEAAAAA90xENER+IjZu+C6NhKKj83zqYCi1HtYCORsV9opylfAAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAYTgRqif/////////8AAAABAAAAAAAAAAAAAAADAYu24wAAAAEAAAAAmlZqRKk9qqEpuyNkDV4LQNI4pBqpW7VR400/bNs2DPUAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAADb1jQf/////////8AAAABAAAAAAAAAAAAAAABAYu3CgAAAAEAAAAAmlZqRKk9qqEpuyNkDV4LQNI4pBqpW7VR400/bNs2DPUAAAABVFhUAAAAAAAVEdwnBqc3YTmCAQ/kcu96/XIgxGlM2ptR5LQ7LK14hAAAAAADYL/lf/////////8AAAABAAAAAAAAAAA="
> xdr_decode = base64.b64decode(xdr)
> unpacker = stellarxdr.Xdr.StellarXDRUnpacker(xdr_decode)
> transaction_meta = unpacker.unpack_TransactionMeta()
> ```
> 
> This part of the code does need to be optimized.

This part : from stellar_base import stellarxdr is not accessible anymore. Could you provide instructions on how it can be don now to decode the tx xdr?

Appreciate it

Brg

**Comment by overcat** [comment]
@AnimusXCASH 
```
import base64
from stellar_sdk import Asset
from stellar_sdk.xdr import Xdr
xdr_str = 'AAAAAVNMVAAAAAAAlA8rrxQ89pexgEv3wGm0Tu9RxisitWx4ONyNc5Xn7qg='
xdr_obj = Xdr.StellarXDRUnpacker(base64.b64decode(xdr_str)).unpack_Asset()
asset = Asset.from_xdr_object(xdr_obj)
```

Please see: 
https://stellar.stackexchange.com/questions/2944/how-to-decode-sellingasset-and-buyingasset-in-python
https://stellar.stackexchange.com/questions/3011/python-sdk-whats-the-difference-between-stellar-base-and-stellar-sdk

**Comment by AnimusXCASH** [comment]
Thank you for assistance but i still cant figure out the way how to get transaction value from xdr envelope. I am building a payment service, where i need to connect Memos with Database accounts and append all necessary data which is read from dictionary. Could you provide example/demo for the payment/transaction? 

Thank you for your time and assistance

**Comment by overcat** [comment]
> Thank you for assistance but i still cant figure out the way how to get transaction value from xdr envelope. I am building a payment service, where i need to connect Memos with Database accounts and append all necessary data which is read from dictionary. Could you provide example/demo for the payment/transaction?
> 
> Thank you for your time and assistance

Demo
```
from stellar_sdk import TransactionEnvelope, Network, TextMemo

# https://horizon.stellar.org/transactions/16a1010d3e3e402f7204733ec2912b300b964d92a9950070f4f19b99ff49a140
envelope_xdr = "AAAAAC4nOKiqX5Qf0+KebfGoR/4yEtrwSpz0TxeIoO23vEFRAAAAZAGlo54AAA4xAAAAAQAAAAAAAAAAAAAAAF5k4nQAAAABAAAACzIwMTkwNzExXzIxAAAAAAEAAAAAAAAAAQAAAAAXZxH7j6D9NrZu+o86lanmSsYzneKWFeVVRId4sLtmxQAAAAFlbmVjAAAAAJqn7errLgH2d2IGqF1gACAfuDHQCFnjT61bJt/qZz30AAAAAHyS3oAAAAAAAAAAAbe8QVEAAABApzGC/JfukU/W9/56wp0qw+qPOuAVsmWcxbluA1MbyQAcG0nd2XGS78XuyrzaIsfjwnMzxlzlRgMs/loWIebnDg=="
te = TransactionEnvelope.from_xdr(envelope_xdr, Network.PUBLIC_NETWORK_PASSPHRASE)
memo = te.transaction.memo
if isinstance(memo, TextMemo):
    print(memo.memo_text)
```

Perhaps asking a question [here](https://stellar.stackexchange.com/) can get more answers.

**Comment by overcat** [comment]
Closed via 3.0.0 release.
